contracts/ibc-reflect/src/contract.rs

```diff
@@ -0,0 +1,205 @@
+    }
+    if channel.version.as_str() != IBC_VERSION {
+        return Err(StdError::generic_err(format!(
+            "Must set version to `{}`",
+            IBC_VERSION
+        )));
+    }
+
+    Ok(())
+}
+
+#[entry_point]
+/// once it's established, we create the reflect contract
+pub fn ibc_channel_connect(
+    deps: DepsMut,
+    _env: Env,
+    channel: IbcChannel,
+) -> StdResult<IbcBasicResponse> {
+    let cfg = config(deps.storage).load()?;
+    let chan_id = channel.endpoint.channel_id;
+    let label = format!("ibc-reflect-{}", &chan_id);
+
+    let payload = to_binary(&ReflectInitMsg {
+        callback_id: Some(chan_id),
+    })?;
+    let msg = WasmMsg::Instantiate {
+        code_id: cfg.reflect_code_id,
+        msg: payload,
+        send: vec![],
+        label: Some(label),
+    };
+
+    Ok(IbcBasicResponse {
+        messages: vec![msg.into()],
+        attributes: vec![],
+    })
+}
+
+#[entry_point]
+/// we do nothing
+/// TODO: remove the account from the lookup?
+pub fn ibc_channel_close(
+    _deps: DepsMut,
+    _env: Env,
+    _channel: IbcChannel,
+) -> StdResult<IbcBasicResponse> {
+    Ok(IbcBasicResponse::default())
+}
+
+// pull this into one function so we can handle all StdErrors once
+fn parse_receipt(deps: DepsMut, packet: IbcPacket) -> StdResult<(HumanAddr, Vec<CosmosMsg>)> {
+    // which local channel did this packet come on
+    let caller = packet.dest.channel_id;
+    // what is the reflect contract here
+    let reflect_addr = accounts(deps.storage).load(caller.as_bytes())?;
+
+    // parse the message we got
+    let msg: PacketMsg = from_binary(&packet.data)?;
+    Ok((reflect_addr, msg.msgs))
+}
+
+#[entry_point]
+/// we look for a the proper reflect contract to relay to and send the message
+/// We cannot return any meaningful response value as we do not know the response value
+/// of execution. We just return ok if we dispatched, error if we failed to dispatch
+pub fn ibc_packet_receive(
+    deps: DepsMut,
+    _env: Env,
+    packet: IbcPacket,
+) -> StdResult<IbcReceiveResponse> {
+    let (reflect_addr, msgs) = match parse_receipt(deps, packet) {
+        Ok(m) => m,
+        Err(_) => {
+            let acknowledgement =
+                to_binary(&AcknowledgementMsg::Err("invalid packet".to_string()))?;
+            return Ok(IbcReceiveResponse {
+                acknowledgement,
+                messages: vec![],
+                attributes: vec![],
+            });
+        }
+    };
+
+    // let them know we're fine
+    let acknowledgement = to_binary(&AcknowledgementMsg::Ok(()))?;
+    // create the message to re-dispatch to the reflect contract
+    let reflect_msg = ReflectHandleMsg::ReflectMsg { msgs };
+    let wasm_msg = WasmMsg::Execute {
+        contract_addr: reflect_addr,
+        msg: to_binary(&reflect_msg)?,
+        send: vec![],
+    };
+    // and we are golden
+    Ok(IbcReceiveResponse {
+        acknowledgement,
+        messages: vec![wasm_msg.into()],
```

Comment:
How can be ensure that only "authorized" messages are passed?

Reply:
When we create a channel, there is exactly one contract/module on the other side that can send packets on it, correct? (Please double-check this - this is key to all my authorization reasoning).

When they connect, we give them a fresh address that is theirs. Any messages they send get reflected by that custom address. Thus the channel -> address lookup.

If sending packets on the channel is not as limited as I believe, I need to revisit this.

Reply:
Note: We really need local `(port_id, channel_id)` pairs for uniqueness.
However, the receiving contract only has one port, so by fixing the channel_id, we fix the pair.